apache2/msc_multipart.c

```diff
@@ -214,23 +214,42 @@ static int multipart_parse_content_disposition(modsec_rec *msr, char *c_d_value)
                 }
                 msr->multipart_filename = decoded_filename;
 
+               
+                if (msr->mpd->mpp->filenameext != NULL) {
+                    msr_log(msr, 4, "Multipart: Warning: Duplicate Content-Disposition filename*: %s",
+                     log_escape_nq(msr->mp, decoded_filename));
+                     return -17;
+                }
+
+                msr->mpd->mpp->filenameext = apr_pstrdup(msr->mp, decoded_filename);
+
+                // The `filename*` RCF 5987 encoded filename always overrides the `filename` parameter in content-disposition header.
+                msr->mpd->mpp->filename = msr->mpd->mpp->filenameext;
+
                 // Make sure to turn of INVALID quoting since RFC 5987 expects quotes in the filename format.
                 msr->mpd->flag_invalid_quoting = 0;
             }
             else
             {
-                decoded_filename = value;
-                validate_quotes(msr, value);
-                msr->multipart_filename = apr_pstrdup(msr->mp, decoded_filename);
-            }
+                // Process the `filename` attribute in the content-disposition header only if `filename*` does not exist.
+                if (msr->mpd->mpp->filenameext == NULL)
+                {
+                    // "name == 'filename'"
+                    decoded_filename = value;
+                    validate_quotes(msr, value);
+                    msr->multipart_filename = apr_pstrdup(msr->mp, decoded_filename);
 
-            if (msr->mpd->mpp->filename != NULL)
-            {
-                msr_log(msr, 4, "Multipart: Warning: Duplicate Content-Disposition filename: %s",
-                        log_escape_nq(msr->mp, decoded_filename));
-                return -15;
+                    if (msr->mpd->mpp->filename != NULL)
```

Comment:
in two cases, 
case 1, filename, filename, filename*  it will print warning duplicate filename
case 2 filename*, filename, filename it won't print out any warning, 
Are we ok with this logging behavior difference?

Reply:
This is a good point. Let me think about it. It does create an inconsistent behavior.

Reply:
The problem is not just with logging, in case 1 it will block the request with a return code of -15 and case 2 it will allow the request to go through.